stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,239 @@
+	// the string "other". If unset, will use the target string as is.
+	TargetAttributeFilter func(string) bool
+	// MethodAttributeFilter is a callback that takes the method string and
+	// returns a bool representing whether to use method as a label value or use
+	// the string "other". If unset, will use the method string as is. This is
+	// used only for generic methods, and not registered methods.
+	MethodAttributeFilter func(string) bool
+}
+
+// DialOption returns a dial option which enables OpenTelemetry instrumentation
+// code for a grpc.ClientConn.
+//
+// Client applications interested in instrumenting their grpc.ClientConn should
+// pass the dial option returned from this function as a dial option to
+// grpc.Dial().
+//
+// For the metrics supported by this instrumentation code, a user needs to
+// specify the client metrics to record in metrics options. A user also needs to
+// provide an implementation of a MeterProvider. If the passed in Meter Provider
+// does not have the view configured for an individual metric turned on, the API
+// call in this component will create a default view for that metric.
+func DialOption(mo MetricsOptions) grpc.DialOption {
+	csh := &clientStatsHandler{mo: mo}
+	csh.initializeMetrics()
+	return joinDialOptions(grpc.WithChainUnaryInterceptor(csh.unaryInterceptor), grpc.WithStreamInterceptor(csh.streamInterceptor), grpc.WithStatsHandler(csh))
+}
+
+// ServerOption returns a server option which enables OpenTelemetry
+// instrumentation code for a grpc.Server.
+//
+// Server applications interested in instrumenting their grpc.Server should pass
+// the server option returned from this function as an argument to
+// grpc.NewServer().
+//
+// For the metrics supported by this instrumentation code, a user needs to
+// specify the client metrics to record in metrics options. A user also needs to
+// provide an implementation of a MeterProvider. If the passed in Meter Provider
+// does not have the view configured for an individual metric turned on, the API
+// call in this component will create a default view for that metric.
+func ServerOption(mo MetricsOptions) grpc.ServerOption {
+	ssh := &serverStatsHandler{mo: mo}
+	ssh.initializeMetrics()
+	return grpc.StatsHandler(ssh)
+}
+
+// callInfo is information pertaining to the lifespan of the RPC client side.
+type callInfo struct {
+	target string
+
+	method string
+}
+
+type callInfoKey struct{}
+
+func setCallInfo(ctx context.Context, ci *callInfo) context.Context {
+	return context.WithValue(ctx, callInfoKey{}, ci)
+}
+
+// getCallInfo returns the callInfo stored in the context, or nil
+// if there isn't one.
+func getCallInfo(ctx context.Context) *callInfo {
+	ci, _ := ctx.Value(callInfoKey{}).(*callInfo)
+	return ci
+}
+
+// rpcInfo is RPC information scoped to the RPC attempt life span client side,
+// and the RPC life span server side.
+type rpcInfo struct {
+	mi *metricsInfo
+}
```

Comment:
This only has one field...is it needed?

Also "callInfo" vs. "rpcInfo" -- what's the difference?

Reply:
This extra layer is needed for when we eventually add tracing (traceInfo). callInfo is the callInfo client side (i.e. scoped to full call, not attempt). I thought about prepending this with client - clientCallInfo to be more clear. Would this help? rpcInfo is scoped to RPC attempt life span client side, and RPC life span server side.

Reply:
If `rpcInfo` is meant to be per-attempt then it should probably be `attemptInfo` instead.

Reply:
It's also used on server, which doesn't have the concept of an attempt, so that language seems wrong to me. Is there anything more specific client side you can think of that maps to a concept server side?